### PR TITLE
return 404 for "not a directory" errors

### LINKED ISSUE #1408: caddy returns 503 if path contains a file
### 1. What version of Caddy are you running (`caddy -version`)?
master branch

### 2. What are you trying to do?
GET request on `existingfile/something`

### 3. What is your entire Caddyfile?
```text
:2015
```

### 4. How did you run Caddy (give the full command and describe the execution environment)?
`caddy`

### 5. What did you expect to see?
Expected a 404 response

### 6. What did you see instead (give full error messages and/or log)?
Got a 503  response

### 7. How can someone who is starting from scratch reproduce this behavior as minimally as possible?

create `foo.html`
run `caddy`
run `curl localhost:2015/foo.html/somepath`

### PR DESCRIPTION
**Pull request by mastercactapus**
Fixes #1408

### CODE REVIEW

**Comment by mholt** [comment]
Thanks! I'm kind of watching https://github.com/golang/go/issues/18974 and now https://github.com/golang/go/issues/18984 to see what happens and when/if it gets fixed in the standard lib. If it does get fixed there, we could remove this hack later, which means we'd want to accompany it with a TODO of some sort.

File: caddyhttp/staticfiles/fileserver.go [hunk 1]
```diff
@@ -40,6 +40,15 @@ func (fs FileServer) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, err
...
 
+func isNotDirErr(err error) bool {
+	perr, ok := err.(*os.PathError)
+	if !ok {
+		return false
+	}
+
+	return perr.Err.Error() == "not a directory"
+}
+
```
Comment (mholt) [inline caddyhttp/staticfiles/fileserver.go]: Can we move this lower in the file to keep the serveFile and ServeHTTP methods close together?

Comment (mholt) [inline caddyhttp/staticfiles/fileserver.go]: Also a godoc on this would be useful.

Comment (mastercactapus) [inline caddyhttp/staticfiles/fileserver.go]: Yeah, I'm actually going to revisit this and update the PR with the fix I'm using/submitting to golang. I'll add the TODO and comments as well -- hopefully sometime today.

**Comment by mastercactapus** [comment]
@mholt I think this is ready to be looked at again when you get a chance. The fix has now been merged into golang/go. I've used that implementation here, and added TODO comments to remove it when the fix is released.

Let me know if you want me to revise anything!

**Comment by mholt** [comment]
Thanks @mastercactapus  - I see they merged a fix into Go. Am I correct in understanding that this patch is temporary, then, until they release a Go version that has your patch?

**Comment by mastercactapus** [comment]
@mholt That is correct. I also tested without this patch against Go master to make sure it was resolved (and how I found an additional case). This is the same code that got merged into Go, adapted of course.

**Comment by mholt** [comment]
@mastercactapus Cool. When that patch is released, will you update Caddy so this patch won't be redundant? If so, I'll merge this.

**Comment by mastercactapus** [comment]
@mholt definitely, I can own it.

**Comment by mholt** [comment]
@mastercactapus Mind if I invite you to be a collaborator? I can't remember if I've done this for you yet, but it doesn't look like I have. This way you could contribute in a branch instead of a fork, and help with reviewing pull requests and issues when you are able.

**Comment by mastercactapus** [comment]
@mholt I don't mind, I'd be happy to

**Pull request merged**